contracts/adapters/swap/astroport/src/contract.rs

```diff
@@ -190,29 +191,37 @@ fn query_simulate_swap_exact_coin_in(
         return Err(ContractError::CoinInDenomMismatch);
     }
 
-    // Get the router contract address
-    let router_contract_address = ROUTER_CONTRACT_ADDRESS.load(deps.storage)?;
-
-    // Get denom out from last swap operation
-    let denom_out = last_op.denom_out.clone();
+    // Iterate through the swap operations, querying the astroport pool contracts to get the coin out
+    // for each swap operation, and then updating the coin out for the next swap operation until the
+    // coin out for the last swap operation is found.
+    let coin_out = swap_operations.iter().try_fold(
+        coin_in,
+        |coin_out, operation| -> Result<_, ContractError> {
+            let res: SimulationResponse = deps.querier.query_wasm_smart(
+                &operation.pool,
+                &PairQueryMsg::Simulation {
+                    offer_asset: Asset {
+                        info: AssetInfo::NativeToken {
+                            denom: coin_out.denom,
+                        },
+                        amount: coin_out.amount,
+                    },
+                    ask_asset_info: None,
+                },
+            )?;
 
-    // Convert the swap operations to astroport swap operations
-    let astroport_swap_operations = swap_operations.into_iter().map(From::from).collect();
+            // Assert the operation does not exceed the max spread limit
+            assert_max_spread(res.return_amount, res.spread_amount)?;
```

Comment:
Does the `?` force this to return early if the assertion is false?

Reply:
Yup, exactly, it can be used after any expression that returns a result, and it'll unwrap the result into the value if it's a value, and pass up the error if it's an error.